feat: implemented bindings to onnxruntime on-device training C api

riccardopinosio opened the pull request:
This PR implements initial support for bindings to the onnxruntime training C api for on-device training.

closes https://github.com/yalue/onnxruntime_go/issues/52

@yalue this is an initial PR, it's not ready yet (need to add comments, and at least a benchmark and test on gpu) but i would value your initial comments and feedback on it.

Some highlights to think about:

- currently onnxruntime_go only supports the creation of tensors where the `Shape` is not null/zero. However, for onnxruntime training I need to be able to create scalar ortValues (tensors of nil dimension) to hold, e.g., the loss from a training epoch (see tests). Currently I added a `Scalar` type which implements `ArbitraryTensor` to do this, but one could also think to modify the creation of tensors to allow for nil shapes (I do like the distinction between scalars and tensors, however).
- I currently have two functions to create a training session: `NewTrainingSession` and `NewTrainingSessionWithOnnxData`. These map respectively to the C `createTrainingSession` and `createTrainingSessionFromBuffer` functions. Originally I only wanted to use the latter and read the models as bytes like the AdvancedSession does, however it appears there is no way to export the final model if the session is not initiated via `createTrainingSession` on the C side, because it needs the paths to the `eval` model. Hence I have both functions now, but `NewTrainingSessionWithOnnxData` cannot be used to save the final model (it can export the checkpoint, however).

How to test the PR: from the root folder do:

```
python test_data/setup_onnxruntime_training.py --onnxruntime $onnxruntime_version --platform $platform
```

This will create a test_data/onnxruntime_training_test folder with the onnxruntime training library and the required assets for training. The assets replicate this example:

https://github.com/microsoft/onnxruntime-training-examples/blob/master/on_device_training/mobile/android/java/train.ipynb

After that, the tests in onnxruntime_training_test.go can be executed.

yalue commented:
Wow!  That's some serious work, I wasn't sure if you were actually planning to follow through so I'm happy to see it.

As for some initial thoughts:
 - I like your addition of the `Scalar` type that supports the `ArbitraryTensor` interface; that seems like the right move to me.  I'd recommend also adding a `func (s *Scalar[T]) Set(value T)` function to set the Scalar's value, since a user currently doesn't have a way to modify the scalar value after creation due to the scalar's `GetData` function not returning the slice.
 - Unfortunately, I was avoiding passing paths to onnxruntime for a particular reason: Windows compatibility.  It's highly obnoxious (IMO), but on Windows, the `ORTCHAR_T` type expects a wide-character string so I expect that casting paths to a `*C.char` will fail on Windows.  I worked around this in the non-training library by never passing paths to onnxruntime, but maybe you can come up with some clever workaround that doesn't rely on a third-party library.  Worst case, we can gate off this function entirely on Windows.
 - When it comes to testing, I'd rather avoid the pre-step of generating new data.  Would it be possible to skip the python script entirely and just train a network that's even simpler?  My suggestion would be mapping a 2xN input tensor -> 1xN output vector where the output vector just contains the sum of each 2 input elements.  This could be generated entirely within the Go testing code, and verified to be within some generous tolerance.
 - In testing code, don't forget to `Destroy()` the session options, sessions, and so on, after creating them.  It's not likely to be a problem in the tests, but it will help the tests serve as best-practices examples if people refer to them when learning to use the library.

Otherwise, this looks pretty great!  Don't worry too much about the windows paths thing.  If you want to gate off the training API on Windows for now, that would probably be fine with me.  Just don't call `C.SetTrainingApi()` in `setup_env_windows.go`, and make sure to include a comment about the reason.

yalue commented:
Oh, another couple comments on tests that I forgot:
 - I'd rather not call `t.Setenv(...)` in the tests.  I'd think the burden should be on the user to set the variable in their own environments rather than assuming its presence in test_data.
 - Preferably start each test with something like this to keep tests from failing on builds without training support:
    ```go
    if !IsTrainingSupported() {
        t.Skipf("Training is not supported on this platform/onnxruntime build.")
    }
    ```

riccardopinosio commented:
Ok cool, a couple of follow ups:

- indeed it might be good to gate this for windows, with the current C api i don't see an easy way to avoid passing paths. It might be that i open an issue with onnxruntime (it's a bit weird that you need to pass paths to be able to save the final model, for instance) and if that is improved we can extend this to windows.
- on the tests, I can indeed make the test without the python script, I assume then I should only commit the testing assets (training, eval, optimizer models) and not commit the onnxruntime_training.so file and the user will have to download the library himself?

yalue commented:
I don't know if the onnxruntime maintainers are likely to want to support utf-8 strings on Windows, since they'll likely want to continue using the default behavior of MSVCRT when building the Windows DLLs.  Additionally, it _should_ be possible for us to convert the utf-8 strings into the wide-character strings when passing paths on Windows, using only the Go standard library.  I personally just wasn't able to figure out how, and decided the workaround was simpler.  Now that there's a use case, it's probably better for me (or someone else) to take responsibility for addressing this issue in Go.

For now, though, I'm fine with simply making this unsupported on Windows.

>I assume then I should only commit the testing assets (training, eval, optimizer models) and not commit the onnxruntime_training.so file and the user will have to download the library himself?
That's what I would prefer, yes.  After merging your change and playing with it myself, I will likely update the main README with a section of additional requirements for people wanting to use the training API.

yalue commented:
To clarify regarding content to add to test_data: while I'd like to avoid adding new onnxruntime libraries or datasets, adding the models along with a script to generate them sounds great, so long as the models are reasonably small.

riccardopinosio commented:
@yalue I've refactor the PR to simplify a few things:

- I now use the same network that you use in the tests for the training test, i.e., the one that calculates the sum and max diff of four numbers. This simplifies the main training test.
- I've done a cleanup of the comments, destruction of resources, etc.
- I've disabled training on windows and added a comment about it

For the benchmarks, would it be helpful to add a benchmark based on the same network as the main training test?

yalue commented:
Sorry for the slow movement on this.  The self-contained test looks great.  I plan to test it myself (hopefully later today), and plan to merge it if I don't see any problems.  Thanks again!

riccardopinosio commented:
no worries! If you merge it, could you keep me as the commit author? Then it would count me as a contributor to the repo which would be nice :)

yalue commented:
Of course!  If I haven't done so with other past contributions it was purely a mistake due to not understanding github's interface well enough.

riccardopinosio commented:
btw do you still think i should add a benchmark, maybe based on doing training with the simple model?

yalue commented:
I mostly use the benchmarks in the existing code as a sanity check that accelerated execution providers actually improve performance.  For now, I'd say don't worry about a benchmark unless you already have one in mind.  It's fine to put it off until a later change if it seems like one will be useful.

yalue commented on onnxruntime_training_go.go:
```diff
@@ -0,0 +1,531 @@
+// initialized with NewTrainingSession and the path to the eval model has been provided.
+func (s *TrainingSession) ExportModel(path string, outputNames []string) error {
+	if s.evalModelPath == nil {
+		return fmt.Errorf(`final model can only be exported if the eval model path is 
```
Preferably splice together this long string using `+`, i.e.,
```go
fmt.Errorf("first part " +
    "second part")
```
since the backtick will include the newline in the error message.

yalue commented on onnxruntime_training_go.go:
```diff
@@ -0,0 +1,531 @@
+func (s *TrainingSession) TrainStep() error {
+	inputLength := C.size_t(len(s.inputs))
+	outputLength := C.size_t(len(s.outputs))
+	status := C.TrainStep(s.ortTrainingSession, inputLength, &s.inputs[0], outputLength, &s.outputs[0])
```
Do you make sure that inputs and outputs are have a non-zero length somewhere?  If not, doing `&s.inputs[0]` will cause a panic if the length is 0, regardless of the `inputLength` arguments.  (I had this same issue in the non-training API.)  Preferably, just ensure that `inputs` and `outputs` have nonzero lengths when creating the `TrainingSession`.  (I apologize if you already check this somewhere and I missed it!)

yalue commented:
In addition to the couple minor comments on the code, I started by trying to test this on a build of onnxruntime that didn't include training support and got some test failures and panics:
```
--- SKIP: TestCUDASession (0.00s)
=== RUN   TestTensorRTSession
    onnxruntime_test.go:1371: Error creating TensorRT provider options; TensorRT execution provider is not enabled in this build.. Your version of the onnxruntime library may not include TensorRT support. Skipping the remainder of this test.
--- SKIP: TestTensorRTSession (0.00s)
=== RUN   TestCoreMLSession
    onnxruntime_test.go:1425: Couldn't enable CoreML: Your platform or onnxruntime library does not support CoreML. This may be due to your system or onnxruntime library version not supporting CoreML.
--- SKIP: TestCoreMLSession (0.00s)
=== RUN   TestDirectMLSession
    onnxruntime_test.go:1457: Couldn't enable DirectML: Specified provider is not supported.. This may be due to your system or onnxruntime library version not supporting DirectML.
--- SKIP: TestDirectMLSession (0.00s)
=== RUN   TestOpenVINOSession
    onnxruntime_test.go:1490: Couldn't enable OpenVINO: /onnxruntime_src/onnxruntime/core/session/provider_bridge_ort.cc:1375 void onnxruntime::ProviderSharedLibrary::Ensure() [ONNXRuntimeError] : 1 : FAIL : Failed to load library libonnxruntime_providers_shared.so with error: libonnxruntime_providers_shared.so: cannot open shared object file: No such file or directory
        . This may be due to your system or onnxruntime library version not supporting OpenVINO.
--- SKIP: TestOpenVINOSession (0.00s)
=== RUN   TestTrainingNotSupported
    onnxruntime_training_test.go:29: Creating training session when onnxruntime lib does not support it should return error.
--- FAIL: TestTrainingNotSupported (0.00s)
=== RUN   TestSessionInfo
    onnxruntime_training_test.go:75: Failed creating training session: Reading checkpoint data failed: open test_data/training_test/checkpoint: no such file or directory
--- FAIL: TestSessionInfo (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x542cd7]

goroutine 16 [running]:
testing.tRunner.func1.2({0x56d860, 0x6b2df0})
	/home/yalue/go/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	/home/yalue/go/src/testing/testing.go:1548 +0x397
panic({0x56d860?, 0x6b2df0?})
	/home/yalue/go/src/runtime/panic.go:914 +0x21f
github.com/yalue/onnxruntime_go.(*TrainingSession).Destroy(0x0)
	/home/yalue/Dropbox/Programming/go/src/github.com/yalue/onnxruntime_go/onnxruntime_training_go.go:170 +0x17
github.com/yalue/onnxruntime_go.TestSessionInfo.func1(0x4cfdea?)
	/home/yalue/Dropbox/Programming/go/src/github.com/yalue/onnxruntime_go/onnxruntime_training_test.go:67 +0x2c
runtime.Goexit()
	/home/yalue/go/src/runtime/panic.go:523 +0x145
testing.(*common).FailNow(0xc00008ad00)
	/home/yalue/go/src/testing/testing.go:999 +0x4a
github.com/yalue/onnxruntime_go.TestSessionInfo(0xc00008ad00)
	/home/yalue/Dropbox/Programming/go/src/github.com/yalue/onnxruntime_go/onnxruntime_training_test.go:76 +0x4b4
testing.tRunner(0xc00008ad00, 0x5a2988)
	/home/yalue/go/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	/home/yalue/go/src/testing/testing.go:1648 +0x3ad
exit status 2
FAIL	github.com/yalue/onnxruntime_go	2.595s
```

I am assuming that `TestTrainingNotSupported` should `PASS` when using an unsupported build, right?  And `TestSessionInfo` should be skipped in this case.

Sorry it took me a while to pull out my Linux system to start testing!

riccardopinosio commented:
Hi, I added checks for the input size and reformat the strings.

The thing that is puzzling me now are those panics when training is not supported, when I test it in my linux env with the library pointing to the latest cpu onnxruntime for x64 linux (without training) the test for TrainingNotSupported passes and the one for InputOutput is skipped. From that log it seems that the IsTrainingSupported() function is returning true on your system, this in turn depends on line 358-369 of the C wrapper. Could you take a look at those lines and see if there's anything I missed?

yalue commented:
Thanks! I'll check it more carefully and get back to you.  Maybe something's extra-wrong with my setup.

yalue commented:
Just a quick update: I'm still looking into it, maybe my build of onnxruntime does support the training API (at least GetTrainingApi seems to return non-NULL), which is surprising to me but my mistake.  However, the thing causing the panic was the lack of the training data.  So, I'll still try to figure out what's messed up on my system, but there are a few comments in the meantime:

 - Would you mind committing the pre-populated `test_data/training_test` directory as well?  This way, people can just pull the repo and run `go test -v` and expect it to work.  This directory seems to be small, only a few KB, so including these .onnx files seems fine to me.
 - I don't know how to initially generate `test_data/training_test/checkpoint`, which is what's currently causing `TestTrainingNotSupported` to fail for me.

I also added a few comments again.

Thanks again!

yalue commented:
Somehow, training does seem to be supported on my onnxruntime build, so my tests are progressing farther than before.  Here is the current output of `go test -v` for me:
```
=== RUN   TestSessionInfo
--- PASS: TestSessionInfo (0.06s)
=== RUN   TestTraining
Epoch {1} Loss {0.128927}
Epoch {11} Loss {0.109867}
Epoch {21} Loss {0.109683}
Epoch {31} Loss {0.109506}
Epoch {41} Loss {0.109350}
Epoch {51} Loss {0.109180}
Epoch {61} Loss {0.109061}
Epoch {71} Loss {0.108956}
Epoch {81} Loss {0.108840}
Epoch {91} Loss {0.108732}
--- FAIL: TestTraining (0.14s)
FAIL
exit status 1
FAIL	github.com/yalue/onnxruntime_go	3.008s
```

So, it's cool that it seems to be working, at least up to a point!

yalue commented on onnxruntime_training_test.go:
```diff
@@ -0,0 +1,328 @@
+		nil,
+		options)
+
+	defer func(session *TrainingSession) {
```
This `defer` to destroy the session needs to be moved after the `if errorSessionCreation != nil` check.  This was what was causing my panic earlier when the `test_data` was missing---the session wasn't created properly (and an error was returned), but the test still panic'd when Destroy was called on the nil session.

riccardopinosio commented:
Thanks, will look at implementing fixes asap

riccardopinosio commented:
ok! I have cleaned the error stuff up and added the test data for the training test, let me know if the tests now run or you encounter more issues

yalue commented:
Once again, sorry for the slowness.  Looking at it now.  The changes look good, but I am now getting a different panic.  I'll look into it a bit more and try to figure out why.

yalue commented:
Alright, I figured out the panic: it was due to `t.Errorf(...)` not causing the test to immediately fail when creating a session, causing `.Destroy()` to be called on a nil session similar to last time.  I was able to fix this pretty easily by replacing `t.Errorf(...)` with `t.Fatalf(...)` in most cases in `onnxruntime_training_test.go`.

However, the tests still failed:
```text
=== RUN   TestTrainingNotSupported
    onnxruntime_training_test.go:30: Creating training session when onnxruntime lib does not support it should return training not supported error.
    onnxruntime_training_test.go:32: Received instead error: inputs must have length greater than zero
--- FAIL: TestTrainingNotSupported (0.00s)
=== RUN   TestSessionInfo
    onnxruntime_training_test.go:68: Failed creating training session with error: inputs must have length greater than zero
--- FAIL: TestSessionInfo (0.00s)
=== RUN   TestTraining
Epoch {1} Loss {0.125085}
Epoch {11} Loss {0.097187}
Epoch {21} Loss {0.062333}
Epoch {31} Loss {0.024307}
Epoch {41} Loss {0.019963}
Epoch {51} Loss {0.018476}
Epoch {61} Loss {0.017160}
Epoch {71} Loss {0.015982}
Epoch {81} Loss {0.014845}
Epoch {91} Loss {0.013867}
    onnxruntime_training_test.go:259: loss deviation too large
--- FAIL: TestTraining (0.09s)
FAIL
exit status 1
FAIL	github.com/yalue/onnxruntime_go	2.323s
```

It looks like this happened because you added the check I requested last time, but still use `nil` when creating a session simply to query information.  This type of problem is why I made `GetInputOutputInfo(...)` into a standalone function rather than a method on `AdvancedSession`.  Though I can see that getting the separate names for training and eval may be a bit more complicated.  I'll leave it to your judgement.  Perhaps the best way is just to go back to what you were doing before, but just adding a `len(inputs) != 0` check prior to `unsafe.Pointer(&inputs[0])` (and similar for outputs).

riccardopinosio commented:
Hi, i think it's better if this api resembles the inference api as much as possible, so I've refactored the `GetInputOutputInfo(...)` as a standalone function. I've also added some fixes to the session creation functions because the eval model is optional so it can actually be zero bytes

yalue commented:
Seems almost perfect, the only test failing now is `TestTrainingNotSupported`, due to it still passing `nil` for both inputs and outputs to `NewTrainingSession`:

```
=== RUN   TestTrainingNotSupported
    onnxruntime_training_test.go:30: Creating training session when onnxruntime lib does not support it should return training not supported error.
    onnxruntime_training_test.go:32: Received instead error: inputs must have length greater than zero
--- FAIL: TestTrainingNotSupported (0.00s)
```

Sorry that it's been taking so long, and that I haven't been very quick at testing these changes.  Recently, I have only been having a couple days a week during which I can work on this project.  The contribution is still appreciated!

yalue commented:
For the record I also tested it on an orin nano, where I built onnxruntime from source and didn't include training support.  The training tests all got skipped as I wanted.  To cover all the bases, I still need to run the tests on Windows, but judging from the source I'm sure that will be fine.

riccardopinosio commented:
no worries, we're all busy! I will fix that test

riccardopinosio commented:
ok, added a small commit that skips that test when the onnx library supports training, cause the whole idea there is to test that if the library _does not_ support training, _then_ the correct error should be returned.

yalue commented:
Alright, all the tests pass on Windows, amd64 Linux, and arm64 Linux.  (Though the only one of those in which the tests weren't skipped was amd64 Linux.)  So I'm happy now!  Feel free to squash if you want, though I don't care about having a chain of commits in there.  I'll merge this when you say you're ready.  Thanks again!

riccardopinosio commented:
ok squashed! Can be merged.
